File: src/main/zhangliang/subdomainvisits/SubdomainVisits.java
package main.zhangliang.subdomainvisits;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

/**
 * https://leetcode.com/problems/subdomain-visit-count/description/
 */
public class SubdomainVisits {

    public static void main(String[] args) {
        System.out.println(new SubdomainVisits().subdomainVisits(new String[] { "9001 discuss.leetcode.com" }));
    }

    /**
     * 52 / 52 test cases passed.Runtime: 28 ms
     * 
     * @param cpdomains
     * @return
     */
    public List<String> subdomainVisits(String[] cpdomains) {
        List<String> result = new ArrayList<>();
        Map<String, Integer> map = new HashMap<>();
        if (cpdomains != null && cpdomains.length > 0) {
            for (String cpdomain : cpdomains) {
                countDomainVisists(cpdomain, map);
            }
        }
        Iterator<Entry<String, Integer>> iterator = map.entrySet().iterator();
        while (iterator.hasNext()) {
            Entry<String, Integer> entry = iterator.next();
            result.add(entry.getValue() + " " + entry.getKey());
        }
        return result;
    }

    private static void countDomainVisists(String cpdomain, Map<String, Integer> map) {
        String[] arr = cpdomain.split(" ");
        Integer count = Integer.valueOf(arr[0]);
        cpdomain = arr[1];
        Integer existedCount = map.get(cpdomain);
        if (existedCount == null) {
            map.put(cpdomain, count);
        } else {
            map.put(cpdomain, count + existedCount);
        }
        while (true) {
            int index = cpdomain.indexOf(".");
            if (index < 0) break;
            cpdomain = cpdomain.substring(index + 1);
            existedCount = map.get(cpdomain);
            if (existedCount == null) {
                map.put(cpdomain, count);
            } else {
                map.put(cpdomain, count + existedCount);
            }
        }
    }
}

File: src/main/houpo/subdomainVisits/subdomainVisits.js

/**
* https://leetcode.com/problems/subdomain-visit-count/description/
You are here! 
Your runtime beats 100.00 % of javascript submissions. * @param {string[]} cpdomains
 * @return {string[]}
 */
var subdomainVisits = function(cpdomains) {
    var map = {};
    for(var j=0;j<cpdomains.length;j++){
        var index = cpdomains[j].indexOf(' ');
        var count = cpdomains[j].substring(0, index) * 1;
        var domainStr = cpdomains[j].substr(index + 1);
        for(var i=0;i<domainStr.length;i++){
            if(i==0 || domainStr[i] == '.'){
                var domain = domainStr.substr(i == 0?0:i+ 1);
                map[domain] = (map[domain]?map[domain] + count:count);
            }
        }
    }
    var reArr = [];
    for(var p in map){
        reArr.push(map[p] + ' ' +p);
    }
    return reArr;
};
console.log(subdomainVisits(["900 google.mail.com", "50 yahoo.com", "1 intel.mail.com", "5 wiki.org"]));
